README
Gophernicus - Copyright (c) 2009-2012 Kim Holviala <kim@holviala.com>

Gophernicus is a modern full-featured (and hopefully) secure gopher
daemon for inetd. It is licensed under the BSD license.

Command line options:
    -h hostname   Change server hostname (FQDN)      [$HOSTNAME]
    -p port       Change server port                 [70]
    -r root       Change gopher root                 [/var/gopher]
    -t type       Change default gopher filetype     [0]
    -g mapfile    Change gophermap file              [gophermap]
    -a tagfile    Change gophertag file              [gophertag]
    -c cgidir     Change CGI script directory        [/cgi-bin/]
    -u userdir    Change users personal gopherspace  [public_gopher]
    -l logfile    Log to Apache-compatible combined format logfile

    -w width      Change default page width          [70]
    -o charset    Change default output charset      [US-ASCII]

    -s seconds    Session timeout in seconds         [1800]
    -i hits       Maximum hits until throttling      [4096]
    -k kbytes     Maximum transfer until throttling  [4194304]

    -f filterdir  Specify directory for output filters
    -e ext=type   Map file extension to gopher filetype

    -R old=new    Rewrite the beginning of a selector
    -D text|file  Set or load server description for caps.txt
    -L text|file  Set or load server location for caps.txt
    -A admin      Set admin email for caps.txt

    -nv           Disable virtual hosting
    -nl           Disable parent directory links
    -nh           Disable menu header (title)
    -nf           Disable menu footer
    -nd           Disable date and filesize in menus
    -nc           Disable file content detection
    -no           Disable charset conversion for output
    -nq           Disable HTTP-style query strings (?query)
    -ns           Disable logging to syslog
    -na           Disable autogenerated caps.txt

    -d            Debug to syslog (not for production use)
    -b            Display the BSD license
    -?            Display this help
 

Setting up a gopher site
========================

After succesfully installing Gophernicus (see INSTALL) you need to set
up the gopher root directory. By default Gophernicus serves documents
from /var/gopher so start by creating that directory and making sure
it's world-readable. Then, simply add files and directories under your
root, fire up a gopher browser (Firefox, Lynx) and open up this URL:

gopher://<HOSTNAME>/   (where <HOSTNAME> is your servers hostname)

That's it, your first gopher site is now up and running. If the links
on the root menu don't work, make sure you are using the -h <HOSTNAME>
parameter in your inetd.conf (with a valid resolveable hostname
instead of <HOSTNAME> - see INSTALL).


Security
========

Gophernicus has been written with high security in mind. There should
be no buffer overflows or memory allocation problems so it should be
safe to run a publicly available gopher server with Gophernicus.

However, the security settings (which are non-changeable) are so strict
that you need to keep one thing in mind. Gophernicus will only serve
world-readable content. Being readable by the server process is not
enough, all files and directories MUST be world-readable or they are
simply hidden from all listings and denied if a client asks for them.


Gophermaps
==========

By default all gopher menus are automatically generated from the
content of the directory being viewed. If you want to have
informational text along with the files, or if you want to completely
replace the generated menu with your own you need to take a look at
gophermaps. See the file README.gophermap for more information.


Gophertags
==========

A gophertag file can be used to virtually rename a directory. Let's
assume that you have a directory called "foo" somewhere - it will
be listed as "foo" in all automatically generated menus. Now if you
create a file foo/gophertag and put the text "bar" into it the menus
will show "bar" but the links will still point to "foo". This is
useful for creating descriptive names for directories without
littering the file system with spaces and weird characters.


Personal gopherspaces
=====================

Gophernicus supports users personal gopherspaces. If a user has
world-readable directory called public_gopher/ under his home, a
request for gopher://<HOSTNAME>/1/~user/ will serve documents from
that directory.


Virtual hosting
===============

Gophernicus supports virtual hosting, or serving more than one logical
domain using the same IP address. Since gopher (RFC1436) doesn't
support virtual hosting this requires some clever (but mostly invisble)
hacks.

To enable virtual hosting create one or more directories under your
gopher root which are named after your domain names. The primary vhost
directory (set with the -h <HOSTNAME> option) must exist or virtual
hosting will be disabled. Then simply add content to the hostname
directories and you're up and running.

Almost.

To make gopher clients work properly with virtual hosting, create a
root gophermap for each of your domains and include the "%" type
character to create a list of all available virtual hosts (see
README.gophermap). The generated virtual host links will be created so
that standard gopher clients will find the correct domain even when
they don't specifically tell the server which host they're trying to
reach.


CGI support
===========

Gophernicus supports most parts of the CGI/1.1 standard. Most standard
CGI variables are set, and some non-standard ones are added.

By default all scripts and binaries under any directory called
/cgi-bin/ are executed as CGI scripts (this includes cgi-bin
directories under users personal gopherspaces). Also, if a gophermap
is marked executable it is also processed as an CGI script.

As with regular files, CGI scripts must be world-executable (and
readable) or they will be ignored. Make sure your CGI script is safe
with ANY user input as poorly coded CGI scripts are the number #1
security problem with publicly open *nix servers.


Output filtering and PHP support
================================

In addition to CGI scripts Gophernicus supports output filtering
scripts. By default output filtering is turned off, but you can turn
it on by using the -f <FILTERDIR> option, creating that directory
and creating one or more scripts in there named by either the file
suffix, or by the gopher filetype char.

If a file is to be served out which matches either the file suffix
script, or the filetype script then instead of simply sending the
file to client the output filter script is executed with the
original file as the first parameter and the output of the script
is then sent to client.

For PHP support install the CLI version of the PHP interpreter and
then symlink (or copy) that binary to the directory specified with
-f option using the destination name "php".

$ ln -s /usr/bin/php5-cli /usr/lib/gophernicus/filters/php

After that all files with the php suffix will be "filtered" through
the PHP command line interpreter. In other words, PHP starts working.
And don't use the CGI version of PHP as it outputs HTTP headers the
gopher protocol doesn't have.


Charset support and conversions
===============================

Gophernicus supports three charsets: US-ASCII, ISO-8859-1 and UTF-8.
All textual input is internally upconverted to UTF-8 and then
downconverted to whatever charset the client is asking for. The
conversion is input autosensing which means that you don't have to
specify your filesystem charset, or the charset of your text files -
it's all detected automatically.

With standard gopher clients this is a bit of a problem as your text
files WILL be converted to 7-bit US-ASCII. This means that all 8-bit
charaters WILL BE LOST. This decision was made because no gopher
client that I tested was reliably cabable of decoding anything else
than pure US-ASCII. If you want to disable the conversion use the
"-no" option, or if you'd like to change the default output charset to
something else than US-ASCII just use for example the "-o ISO-8859-1"
option.


Selector rewriting
==================

Selector rewriting lets you rewrite parts of the selector on the fly.
Well, not parts, but really just the start of it. And the rewrite
enging here is nothing like Apache's mod_rewrite as I was too lazy
to integrate any regex libraries... So, all it does is rewrite a
fixed string at the start of the selector to something else. This
will let you move your directories around while making sure that
existing deeplinks still work.

Examples:

  -R "/~user=/~luser"
  -R "/old-dir=/new-dir"


Session tracking and statistics
===============================

To enable virtual hosting with gopher (RFC1436) clients Gophernicus
tracks users and their session. As a side effect of that session
tracking, Gophernicus has simple throttling controls to keep nasty
users from killing your precious 120MHz PPC 604e server from dying
under the load. The throttling defaults are high enough that normal
human users will never hit the limits, but it's possible (and mostly
preferrable) that a badly behaving crawling agent will be throttled.

The current sessions and other real-time status data can be viewed
by opening the URL gopher://<HOSTNAME>/0/server-status . This status
view has been modeled after the Apache server-status which means
that it's possible to integrate Gophernicus into existing server
monitoring systems. To ease up such integrations, Gophernicus
supports HTTP requests of the server-status page using an URL like
http://<HOSTNAME>:70/server-status?auto




ChangeLog
2012-05-04  Kim Holviala  <kim@holviala.com>

	* Released version 1.2 (finally...)
	* On the fly selector rewriting (like Apache mod_rewrite)
	* Gophertags are properly converted to output charset
	* Server admin email can be specified for caps.txt
	* Server location for caps.txt
	* Split options.c from gophernicus.c
	* Removed support for Gopher over HTTP proxies
	* Hack to make "gopher example.com" work
	  (UMN gopher client assumes gopher+ which we don't support)


2010-12-01  Kim Holviala  <kim@holviala.com>

	* Released version 1.1
	* Content-based detection of gif/png/jpg/ps/pdf/html/gz
	* Gophermap virtual host list (%) only lists FQDNs
	* Serve out caps.txt from a file if it exists
	* Option -na disables autogenerated caps.txt
	* Changed option -l to -b (show BSD license)
	* Changed option -m to -l (log to file)
	* README and LICENSE weren't zero-terminated strings (duh)
	* Querying /server-status no longer updates statistics
	  (because Munin statistics collection was being throttled)
	* Internal charset variables are now enums and not strings


2010-10-05  Kim Holviala  <kim@holviala.com>

	* Released version 1.0
	* Support for caps.txt as suggested by Cameron Kaiser
	* Support for gophertags (lifted from Bucktooth)
	* HTTP requests are redirected to a public gopher proxy
	* Allow directories named "gophermap" and "gophertag"
	  (only files are special)
	* Removed duplicate call to strniconv()
	* A few ENABLE_STRICT_RFC1436 additions
	* Much more descriptive error logging
	* Self references /./ are removed from request
	* Apache-compatible file logging in combined log format (-m)
	* FIFOs in directories no longer crash the server (duh!)
	* Support for Mac OS X (tested with 10.5 on Intel)
	* Reworked Makefile with platform-specific build targets
	* Fixed a call to dirname(path) (don't assume it modifies path)
        * Makefile installer now supports xinetd and launchd
	* Makefile installer installs default /var/gopher/gophermap and
	  links the document directory as /docs/
	* Added SERVER_ARCH, SERVER_VERSION and SERVER_DESCRIPTION to
	  CGI env variables
	* Fixed a segfault where shm was used uninitialized (duh!)
	* Generate native Debian/Ubuntu package with "make deb"
	* platform() now tries to figure out Linux distribution
	* platform() knows about Linux ARM boards
	* Added argument -nm (No shared Memory) for debugging
	* Fixed AIX make glitch when compiling bin2c (bug in make?)
	* Changed the name of the project from "Gophernicus Server"
	  to plain "Gophernicus"
	* Menus without footer (-nf) were missing the dot at the end
	* Footer message is now right-aligned
	* server-status and caps.txt requests now update sessions
	* Compile-time option ENABLE_AUTOHIDING hides manually listed
	  resources from generated menus (to prevent double listings)


2010-07-03  Kim Holviala  <kim@holviala.com>

	* Released version 0.9
	* Added option to disable HTTP-style query strings
	* Fixed a regression where some binary files were served out
	  as text (which broke them)
	* BinHex files are now mapped to filetype 4
	* Disabling vhosting disables sessions (mostly)


2010-04-30  Kim Holviala  <kim@holviala.com>

	* Released version 0.8
	* Security bug in hURL handler fixed
	* Error pages are now correctly generated for type 'h'
	* Menu error page more compatible with clients
	* New macro sstrncmp() compares without explicit sizeof()
	* Removed all traces of gopher++ protocol (extra headers)
	  as it just didn't work with older (circa '92) servers
	* Added option to disable automatic menu headers (titles)
	* Do a chdir() to the resource dir before doing anything
	* Double-slashes were slashed in QUERY_STRING by accident
	* Relative links to external hosts work properly in gophermap
	* Gophermaps can include other gophermaps with =/path/to/file
	* Executable gophermaps are parsed just like static ones


2010-04-13  Kim Holviala  <kim@holviala.com>

	* Released version 0.7
	* This release is feature complete, no new features in sight
	* Support for NetBSD (a typo prevented building - duh)
	* Replaced the poorly-working scandir() with opendir/qsort
	* Directory listings (menus) are limited to 1024 entries
	* Changed the filetype of movies from "v" to ";" (which sucks)
	* Fixed a compatibility issue with bucktooth gophermaps
	* Finally wrote decent documentation (README and INSTALL)


2010-04-11  Kim Holviala  <kim@holviala.com>

	* Released version 0.6
	* platform() results are kept in shared memory
	* Support for AIX 5.1 and newer
	* IPv4-in-IPv6 prefix ::ffff: is removed from remote_addr
	* Replaced install(1) with the install-sh script
	* /server-status CPULoad can be parsed from /usr/bin/uptime
	* Replaced text2c/hexdump with bin2c.c (less dependencies)
	* Files with extension .q (type 7 query) are considered CGIs
	* Filetype handling completely rewritten
	* Configurable gopher filetypes using the "-e ext=X" argument
	* Per-directory filetype overrides in gophermaps with ":ext=X"
	* Refuse to serve out gophermaps (why didn't I catch this before?)
	* Cleaned up main()
	* Output filters - run files through an external program (php!)


2010-04-05  Kim Holviala  <kim@holviala.com>

	* Released version 0.5
	* Code tested to work on 32-bit Linux/armv5tel
	* Filetype 7 query errors are now handled properly
	* HTTP-style query string overrides type 7 query
	* Protocol detection (0/+/++) works properly
	* gopher++ extra headers parsed correctly
	* gopher++ works ok with a patched NSCA Mosaic!
	* ISO-8859-1 (Latin-1) output
	* Full UTF-8 output support (without widechars)
	* Gophermaps are converted to output charset
	* All type 0 output is converted to output charset
	* All charset conversions can be disabled with option -no
	* !Titles in gophermaps are converted to gopher menu titles
	* Automatically generates gopher title resources for menus
	* Compile-time option to strictly adhere to RFC 1436
	* Compile-time option to disable all gopher++ support


2010-03-29  Kim Holviala  <kim@holviala.com>


gophernicus.h
/*
 * Gophernicus - Copyright (c) 2009-2010 Kim Holviala <kim@holviala.com>
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef _GOPHERNICUS_H
#define _GOPHERNICUS_H


/*
 * Features
 */
#undef  ENABLE_STRICT_RFC1436	/* Follow RFC1436 to the letter */
#undef  ENABLE_AUTOHIDING	/* Hide manually listed resources from generated menus*/


/*
 * Platform configuration
 */

/* Defaults should fit standard POSIX systems */
#define HAVE_IPv4		/* IPv4 should work anywhere */
#define HAVE_IPv6		/* Requires modern POSIX */
#define HAVE_PASSWD		/* For systems with passwd-like userdb */
#define PASSWD_MIN_UID 100	/* Minimum allowed UID for ~userdirs */
#define HAVE_LOCALES		/* setlocale() and friends */
#define HAVE_SHMEM		/* Shared memory support */
#define HAVE_UNAME		/* uname() */
#define HAVE_POPEN		/* popen() */
#undef  HAVE_STRLCPY		/* strlcpy() from OpenBSD */
#undef  HAVE_SENDFILE		/* sendfile() in Linux & others */

/* Linux */
#ifdef __linux
#undef PASSWD_MIN_UID
#define PASSWD_MIN_UID 500
#define HAVE_SENDFILE
#endif

/* Add other OS-specific defines here */

/*
 * Include headers
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <string.h>
#include <libgen.h>
#include <time.h>
#include <syslog.h>
#include <errno.h>
#include <pwd.h>
#include <limits.h>
#include <libgen.h>

#ifdef HAVE_SENDFILE
#include <sys/sendfile.h>
#include <fcntl.h>
#endif

#ifdef HAVE_LOCALES
#include <locale.h>
#endif

#ifdef HAVE_SHMEM
#include <sys/ipc.h>
#include <sys/shm.h>
#else
#define shm_state void
#endif

#if defined(HAVE_IPv4) || defined(HAVE_IPv6)
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#endif

#ifdef HAVE_UNAME
#include <sys/utsname.h>
#endif


/*
 * Compile-time configuration
 */

/* Common stuff */
#define CRLF		"\r\n"
#define EMPTY		""
#define PARENT		".."
#define ROOT		"/"

#define FALSE		0
#define TRUE		1

#define QUIT		1
#define OK		0
#define ERROR		-1

#define MATCH		0

/* Gopher filetypes */
#define TYPE_TEXT	'0'
#define TYPE_MENU	'1'
#define TYPE_ERROR	'3'
#define TYPE_QUERY	'7'
#define TYPE_BINARY	'9'
#define TYPE_GIF	'g'
#define TYPE_HTML	'h'
#define TYPE_INFO	'i'
#define TYPE_IMAGE	'I'
#define TYPE_TITLE	'!'

/* HTTP protocol stuff for logging */
#define HTTP_OK		200
#define HTTP_404	404
#define HTTP_DATE	"%d/%b/%Y:%T %z"
#define HTTP_USERAGENT	"Unknown gopher client"

/* Defaults for settings */
#define DEFAULT_ROOT	"/var/gopher"
#define DEFAULT_HOST	"localhost"
#define DEFAULT_PORT	70
#define DEFAULT_TYPE	TYPE_TEXT
#define DEFAULT_MAP	"gophermap"
#define DEFAULT_TAG	"gophertag"
#define DEFAULT_CGI	"/cgi-bin/"
#define DEFAULT_USERDIR	"public_gopher"
#define DEFAULT_PROXY	"http://gopherproxy.org/"
#define DEFAULT_ADDR	"unknown"
#define DEFAULT_WIDTH	70
#define DEFAULT_CHARSET	"US-ASCII"
#define MIN_WIDTH	33
#define MAX_WIDTH	200

/* Session defaults */
#define DEFAULT_SESSION_TIMEOUT		1800
#define DEFAULT_SESSION_MAX_KBYTES	4194304
#define DEFAULT_SESSION_MAX_HITS	4096

/* Dummy values for gopher protocol */
#define DUMMY_SELECTOR	"null"
#define DUMMY_HOST	"null.host\t1"

/* Safe $PATH for exec() */
#define SAFE_PATH	"/usr/bin:/bin"

/* Special requests */
#define SERVER_STATUS	"/server-status"
#define CAPS_TXT	"/caps.txt"

/* Error messages */
#define ERR_ACCESS	"Access denied!"
#define ERR_NOTFOUND	"File or directory not found!"

#define ERROR_HOST	"error.host\t1"
#define ERROR_PREFIX	"Error: "

/* String formats */
#define SERVER_SOFTWARE	"Gophernicus"
#define SERVER_PROTOCOL	"RFC1436"
#define SERVER_SOFTWARE_FULL SERVER_SOFTWARE "/" VERSION " (%s)"
#define HEADER_FORMAT	"[%s]"
#define FOOTER_FORMAT	"Gophered by Gophernicus/" VERSION " on %s"

#define UNITS		"KB", "MB", "GB", "TB", "PB", NULL
#define DATE_FORMAT	"%Y-%b-%d %H:%M"	/* See man 3 strftime */
#define DATE_WIDTH	17
#define DATE_LOCALE 	"POSIX"

#define USERDIR_FORMAT	"~%s", pwd->pw_name	/* See man 3 getpwent */
#define VHOST_FORMAT	"gopher://%s/"

/* ISO-8859-1 to US-ASCII look-alike conversion table */
#define ASCII \
	"E?,f..++^%S<??Z?" \
	"?''\"\"*--~?s>??zY" \
	" !c_*Y|$\"C?<?-R-" \
	"??23'u?*,1?>????" \
	"AAAAAAACEEEEIIII" \
	"DNOOOOO*OUUUUYTB" \
	"aaaaaaaceeeeiiii" \
	"dnooooo/ouuuuyty"

#define UNKNOWN '?'

/* Sizes & maximums */
#define BUFSIZE		1024	/* Default size for string buffers */
#define MAX_HIDDEN	32	/* Maximum number of hidden files */
#define MAX_FILETYPES	128	/* Maximum number of suffix to filetype mappings */
#define MAX_FILTERS	16	/* Maximum number of file filters */
#define MAX_SDIRENT	1024	/* Maximum number of files per directory to handle */

/* Struct for file suffix -> gopher filetype mapping */
typedef struct {
	char suffix[15];
	char type;
} ftype;

/* Struct for keeping the current options & state */
typedef struct {

	/* Request */
	char req_selector[BUFSIZE];
	char req_realpath[BUFSIZE];
	char req_query_string[BUFSIZE];
	char req_referrer[BUFSIZE];
	char req_remote_addr[64];
	char req_filetype;
	off_t req_filesize;

	/* Output */
	int  out_width;
	char out_charset[16];

	/* Settings */
	char server_description[64];
	char server_platform[64];
	char server_root[256];
	char server_host_default[64];
	char server_host[64];
	int  server_port;

	char default_filetype;
	char map_file[64];
	char tag_file[64];
	char cgi_file[64];
	char user_dir[64];
	char log_file[256];
	char gopher_proxy[128];

	char hidden[MAX_HIDDEN][256];
	int hidden_count;

	ftype filetype[MAX_FILETYPES];
	int filetype_count;
	char filter_dir[64];

	/* Session */
	int session_timeout;
	int session_max_kbytes;
	int session_max_hits;

	/* Feature options */
	char opt_parent;
	char opt_header;
	char opt_footer;
	char opt_date;
	char opt_syslog;
	char opt_magic;
	char opt_iconv;
	char opt_vhost;
	char opt_query;
	char opt_shm;
	char debug;
} state;

/* Shared memory for session & accounting data */
#ifdef HAVE_SHMEM

#define SHM_KEY		0xbeeb0006	/* Unique identifier + struct version */
#define SHM_MODE	0600		/* Access mode for the shared memory */
#define SHM_SESSIONS	256		/* Max amount of user sessions to track */

typedef struct {
	long hits;
	long kbytes;

	time_t req_atime;
	char req_selector[128];
	char req_remote_addr[64];
	char req_filetype;

	char server_host[64];
	int  server_port;
} shm_session;

typedef struct {
	time_t start_time;
	long hits;
	long kbytes;
	char server_platform[64];
	char server_description[64];
	shm_session session[SHM_SESSIONS];
} shm_state;

#endif

/* Struct for directory sorting */
typedef struct {
	char	name[128];	/* Should be 256 but we're saving stack space */
	mode_t	mode;
	uid_t	uid;
	gid_t	gid;
	off_t	size;
	time_t	mtime;
} sdirent;


/* File suffix to gopher filetype mappings */
#define FILETYPES \
	"txt","0","pl","0","py","0","sh","0","tcl","0","c","0","cpp","0", "h","0","log","0", \
	"conf","0","php","0","php3","0", \
	"map","1","menu","1", \
	"hqx","4", \
	"Z","5","gz","5","tgz","5","tar","5","zip","5","bz2","5","rar","5","sea","5", \
	"q","7","qry","7", \
	"iso","9","so","9","o","9","rtf","9","xls","9","doc","9","ppt","9", \
	"xlsx","9","docx","9","pptx","9","ttf","9","bin","9", \
	"ics","c","ical","c", \
	"gif","g", \
	"html","h","htm","h","xhtml","h","css","h","swf","h","rdf","h","rss","h","xml","h", \
	"jpg","I","jpeg","I","png","I","bmp","I","svg","I","tif","I","tiff","I", \
	"ico","I","xbm","I","xpm","I","pcx","I", \
	"mbox","M", \
	"pdf","p","ps","p", \
	"mp3","s","wav","s","mid","s","wma","s","flac","s","ogg","s","aiff","s","aac","s", \
	"avi",";","mp4",";","mpg",";","mov",";","qt",";","asf",";","mpv",";", \
	NULL, NULL

/*
 * Useful macros
 */
#define strclear(str) str[0] = '\0';
#define sstrlcpy(dest, src) strlcpy(dest, src, sizeof(dest))
#define sstrlcat(dest, src) strlcat(dest, src, sizeof(dest))
#define sstrncmp(s1, s2) strncmp(s1, s2, sizeof(s2) - 1)
#define sstrniconv(charset, out, in) strniconv(charset, out, in, sizeof(out))


/*
 * Include generated headers
 */
#include "functions.h"
#include "files.h"

#endif


